graftm/taxonomy.py

```diff
@@ -0,0 +1,176 @@
+# You should have received a copy of the GNU General Public License           #
+# along with this program. If not, see <http://www.gnu.org/licenses/>.        #
+#                                                                             #
+###############################################################################
+__author__ = "Joel Boyd, Ben Woodcroft"
+__copyright__ = "Copyright 2015"
+__credits__ = ["Joel Boyd"]
+__license__ = "GPL3"
+__version__ = "0.0.1"
+__maintainer__ = "Joel Boyd"
+__email__ = "joel.boyd near uq.net.au"
+__status__ = "Development"
+###############################################################################
+# System imports
+import logging 
+
+###############################################################################
+###############################################################################
+###############################################################################
+
+class MalformedGreenGenesFormattedFile(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class ProgrammingError(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class MissingTaxonomy(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class Taxonomy:
+
+    TAB='\t'
+    COM=','
+    GG_SEP_0='; '
+    GG_SEP_1=';'
+    
+    def __init__(self):
+        self.taxonomy = {}
+        
+    def _ammend_taxonomy_gaps(self):
+        '''
+        Checks that each value in the self.taxonomy dict has consistent
+        length. 
+        '''
+        try:
+            self.max = max([len(x) for x in self.taxonomy.values()])
+            for taxa_id, taxa_taxonomy_list in self.taxonomy.iteritems():
+                
+                if len(taxa_taxonomy_list) == self.max:
+                    continue
+                else:
+                    ammended_split_taxonomy = \
+                                taxa_taxonomy_list + \
+                                [''] * (self.max - len(taxa_taxonomy_list))
+                self.taxonomy[taxa_id] = ammended_split_taxonomy
+        except:
+            logging.error("Controlled exit from taxonomy.Taxonomy \
+ _ammend_taxonomy_gaps")    
+            ProgrammingError("Error when trying to interpret taxonomy input")
+    
+    def check(self, taxa_id):
+        '''
+        Check whether an id is within the taxonomy description.
+        
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        '''
+        if taxa_id in self.id_set:
+            return True
+        else:
+            return False
+        
+    def entry_ids(self):
+        '''
+        Return a list of all taxonomy entries within the self.taxnomy dict
+        '''
+        return self.taxonomy.keys()
+    
+    def rank(self, taxa_id, rank):
```

Comment:
To me this method name reads like "get me the rank of Acidobacteria". I wonder if instead you can just run `taxonomy[taxon][rank]` instead - this method is not complex enough to exist imo